File: Src/ECScape.Core/Systems/NpcSystem.cs
using ECScape.Core.Components;
using ECScape.Core.Engine;
using ECScape.Core.Entities;
using System.Collections.Concurrent;

namespace ECScape.Core.Systems;

internal sealed class NpcSystem : ISystem
{
    private static readonly Random random = new();

    private readonly ConcurrentDictionary<Entity, DateTime> lastUpdateTimes = new();
    private static readonly TimeSpan updateInterval = TimeSpan.FromMilliseconds(500);

    public void Update(World world, float deltaTime)
    {
        Parallel.ForEach(world.Entities
            .Where(x => x.HasComponent<Npc>() && x.HasComponent<Velocity>()), Update);
    }

    private void Update(Entity entity)
    {
        if (!lastUpdateTimes.TryGetValue(entity, out DateTime lastUpdate))
        {
            lastUpdate = DateTime.MinValue;
            lastUpdateTimes[entity] = lastUpdate;
        }

        if (DateTime.Now - lastUpdate < updateInterval)
        {
            return;
        }
        var velocity = entity.GetRequiredComponent<Velocity>();

        lastUpdateTimes[entity] = DateTime.Now;
        var horizontal = random.Next(-200, 200);
        var vertical = random.Next(-10, 200);
        vertical = Math.Abs(vertical) < 2 ? 0 : vertical;

        velocity.Y = vertical;
        velocity.X = horizontal;
    }
}

File: Web/src/systems/NpcSystem.js
import ISystem from './ISystem.js';
import { PlayerControllable, Velocity, Position, Npc, DamagesPlayer } from '../components/Components.js';

class NpcSystem extends ISystem {
    constructor() {
        super();
        this.lastUpdateTimes = new Map();
        this.updateInterval = 500;
    }

    update(world, deltaTime) {
        world.getEntitiesWith(Npc, Velocity, Position).forEach(e => this.updateEntity(world, e));
    }

    updateEntity(world, entity) {
        const lastUpdate = this.lastUpdateTimes.get(entity) || 0;
        if (Date.now() - lastUpdate < this.updateInterval) return;

        const playerPosition = world.getEntityWith(PlayerControllable)?.getComponent(Position);
        const velocity = entity.getRequiredComponent(Velocity);
        const position = entity.getRequiredComponent(Position);
        const damagesPlayer = entity.getComponent(DamagesPlayer);

        // Bias to move towards the player     
        let left = -1800;
        let right = 1800;
        if (damagesPlayer) {
            const distanceToPlayer = playerPosition ?
                Math.min(Math.max((playerPosition.Left - position.Left) / 40, -1.0), 1.0) : 0;
            const bias = distanceToPlayer * 150.0;
            left += bias;
            right += bias;
        }

        this.updateInterval = Math.random() * 500;
        this.lastUpdateTimes.set(entity, Date.now());
        const horizontal = Math.floor(left + Math.random() * (right - left));
        let vertical = Math.floor(-3800 + Math.random() * 7800);
        vertical = Math.abs(vertical) < 2 ? 0 : vertical;

        velocity.Y = vertical;
        velocity.X = horizontal;
    }
}

export default NpcSystem;